File: 22_专题/离线查询/根号分治/RangeModRangeSum.ts
/* eslint-disable no-inner-declarations */

class RangeModRangeSum {
  private readonly _nums: number[]
  private readonly _belong: Uint32Array
  private readonly _blockStart: Uint32Array
  private readonly _blockEnd: Uint32Array
  private readonly _blockSum: number[]
  private readonly _blockMax: number[]

  /**
   * @param arr 非负整数数组.
   */
  constructor(arr: ArrayLike<number>, blockSize = 60) {
    const n = arr.length
    const copy = Array(n)
    for (let i = 0; i < n; ++i) copy[i] = arr[i]
    const belong = new Uint32Array(n)
    for (let i = 0; i < n; ++i) belong[i] = (i / blockSize) | 0
    const blockCount = (1 + n / blockSize) | 0
    const blockStart = new Uint32Array(blockCount)
    const blockEnd = new Uint32Array(blockCount)
    for (let i = 0; i < blockCount; ++i) {
      blockStart[i] = i * blockSize
      blockEnd[i] = Math.min((i + 1) * blockSize, n)
    }
    const blockSum = Array(blockCount)
    const blockMax = Array(blockCount)
    this._nums = copy
    this._belong = belong
    this._blockStart = blockStart
    this._blockEnd = blockEnd
    this._blockSum = blockSum
    this._blockMax = blockMax
    for (let bid = 0; bid < blockCount; bid++) this._rebuild(bid)
  }

  set(index: number, value: number): void {
    if (index < 0 || index >= this._nums.length) return
    const pre = this._nums[index]
    if (pre === value) return
    this._nums[index] = value
    this._rebuild(this._belong[index])
  }

  /**
   * 区间`[start, end)`模`mod`.
   */
  update(start: number, end: number, mod: number): void {
    if (start < 0) start = 0
    if (end > this._nums.length) end = this._nums.length
    if (start >= end) return
    const bid1 = this._belong[start]
    const bid2 = this._belong[end - 1]
    if (bid1 === bid2) {
      for (let i = start; i < end; ++i) this._nums[i] %= mod
      this._rebuild(bid1)
    } else {
      for (let i = start; i < this._blockEnd[bid1]; ++i) this._nums[i] %= mod
      this._rebuild(bid1)
      for (let bid = bid1 + 1; bid < bid2; ++bid) {
        if (this._blockMax[bid] < mod) continue
        for (let i = this._blockStart[bid]; i < this._blockEnd[bid]; ++i) this._nums[i] %= mod
        this._rebuild(bid)
      }
      for (let i = this._blockStart[bid2]; i < end; ++i) this._nums[i] %= mod
      this._rebuild(bid2)
    }
  }

  /**
   * 查询区间`[start, end)`的和.
   */
  query(start: number, end: number): number {
    if (start < 0) start = 0
    if (end > this._nums.length) end = this._nums.length
    if (start >= end) return 0
    const bid1 = this._belong[start]
    const bid2 = this._belong[end - 1]
    let res = 0
    if (bid1 === bid2) {
      for (let i = start; i < end; ++i) res += this._nums[i]
    } else {
      for (let i = start; i < this._blockEnd[bid1]; ++i) res += this._nums[i]
      for (let bid = bid1 + 1; bid < bid2; ++bid) res += this._blockSum[bid]
      for (let i = this._blockStart[bid2]; i < end; ++i) res += this._nums[i]
    }
    return res
  }

  private _rebuild(bid: number): void {
    this._blockSum[bid] = 0
    this._blockMax[bid] = 0
    for (let i = this._blockStart[bid]; i < this._blockEnd[bid]; ++i) {
      this._blockSum[bid] += this._nums[i]
      this._blockMax[bid] = Math.max(this._blockMax[bid], this._nums[i])
    }
  }
}

export { RangeModRangeSum }

if (require.main === module) {
  // check()
  testTime()
  function testTime(): void {
    const n = 1e5
    const arr = Array(n)
    for (let i = 0; i < n; ++i) arr[i] = (Math.random() * 1e9) | 0
    console.time('RangeModRangeSum')
    const rangeModRangeSum = new RangeModRangeSum(arr)
    for (let i = 0; i < 1e5; ++i) {
      rangeModRangeSum.set(i, i)
      rangeModRangeSum.update(0, i, i)
      rangeModRangeSum.query(0, n)
    }
    console.timeEnd('RangeModRangeSum')
  }

  function check(): void {
    class Mocker {
      private readonly _arr: number[]
      constructor(arr: number[]) {
        this._arr = arr.slice()
      }

      set(index: number, value: number): void {
        this._arr[index] = value
      }

      update(start: number, end: number, mod: number): void {
        for (let i = start; i < end; ++i) this._arr[i] %= mod
      }

      query(start: number, end: number): number {
        let res = 0
        for (let i = start; i < end; ++i) res += this._arr[i]
        return res
      }
    }

    const n = 1e5
    const arr = Array(n)
    for (let i = 0; i < n; ++i) arr[i] = (Math.random() * 1e9) | 0

    const mocker = new Mocker(arr)
    const rangeModRangeSum = new RangeModRangeSum(arr)
    for (let i = 0; i < 1e5; ++i) {
      const op = (Math.random() * 3) | 0
      const start = (Math.random() * n) | 0
      const end = (Math.random() * n) | 0
      const mod = (Math.random() * 1e9) | 0
      if (op === 0) {
        mocker.set(start, mod)
        rangeModRangeSum.set(start, mod)
      } else if (op === 1) {
        mocker.update(start, end, mod)
        rangeModRangeSum.update(start, end, mod)
      } else {
        const ans = mocker.query(start, end)
        const res = rangeModRangeSum.query(start, end)
        if (ans !== res) {
          console.error('oops')
          console.log(arr)
          console.log(start, end, mod)
          console.log(ans, res)
          throw new Error()
        }
      }
    }
    console.log('pass')
  }
}

File: 22_专题/离线查询/根号分治/RangeModRangeSum.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// https://www.luogu.com.cn/problem/CF438D
func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	nums := make([]int, n)
	for i := range nums {
		fmt.Fscan(in, &nums[i])
	}

	R := NewRangeModRangeSum(nums, 30)
	for i := 0; i < q; i++ {
		var op int
		fmt.Fscan(in, &op)
		if op == 1 {
			var start, end int
			fmt.Fscan(in, &start, &end)
			start--
			fmt.Fprintln(out, R.Query(start, end))
		} else if op == 2 {
			var start, end, mod int
			fmt.Fscan(in, &start, &end, &mod)
			start--
			R.Update(start, end, mod)
		} else {
			var pos, value int
			fmt.Fscan(in, &pos, &value)
			pos--
			R.Set(pos, value)
		}
	}

}

type RangeModRangeSum struct {
	nums       []int
	belong     []int
	blockStart []int
	blockEnd   []int
	blockSum   []int
	blockMax   []int
}

// blockSize = 30 (1e5时较快)
func NewRangeModRangeSum(nums []int, blockSize int) *RangeModRangeSum {
	nums = append(nums[:0:0], nums...)
	block := UseBlock(len(nums), blockSize)
	belong, blockStart, blockEnd, blockCount := block.belong, block.blockStart, block.blockEnd, block.blockCount
	blockSum := make([]int, blockCount)
	blockMax := make([]int, blockCount) // <mod 则跳过整块取模
	res := &RangeModRangeSum{
		nums:       nums,
		belong:     belong,
		blockStart: blockStart,
		blockEnd:   blockEnd,
		blockSum:   blockSum,
		blockMax:   blockMax,
	}
	for bid := 0; bid < blockCount; bid++ {
		res.rebuild(bid)
	}
	return res
}

// 0 start end : 查询区间[start,end)的和.
func (r *RangeModRangeSum) Query(start, end int) int {
	if start < 0 {
		start = 0
	}
	if end > len(r.nums) {
		end = len(r.nums)
	}
	if start >= end {
		return 0
	}
	bid1, bid2 := r.belong[start], r.belong[end-1]
	sum := 0
	if bid1 == bid2 {
		for i := start; i < end; i++ {
			sum += r.nums[i]
		}
	} else {
		for i := start; i < r.blockEnd[bid1]; i++ {
			sum += r.nums[i]
		}
		for i := bid1 + 1; i < bid2; i++ {
			sum += r.blockSum[i]
		}
		for i := r.blockStart[bid2]; i < end; i++ {
			sum += r.nums[i]
		}
	}
	return sum
}

// 将区间[start,end)所有数模mod.
func (r *RangeModRangeSum) Update(start, end, mod int) {
	if start < 0 {
		start = 0
	}
	if end > len(r.nums) {
		end = len(r.nums)
	}
	if start >= end {
		return
	}
	bid1, bid2 := r.belong[start], r.belong[end-1]
	if bid1 == bid2 {
		for i := start; i < end; i++ {
			r.nums[i] %= mod
		}
		r.rebuild(bid1)
	} else {
		for i := start; i < r.blockEnd[bid1]; i++ {
			r.nums[i] %= mod
		}
		r.rebuild(bid1)
		for i := bid1 + 1; i < bid2; i++ {
			if r.blockMax[i] < mod {
				continue
			}
			for j := r.blockStart[i]; j < r.blockEnd[i]; j++ {
				r.nums[j] %= mod
			}
			r.rebuild(i)
		}
		for i := r.blockStart[bid2]; i < end; i++ {
			r.nums[i] %= mod
		}
		r.rebuild(bid2)
	}
}

// 单点修改 nums[pos] = value.
func (r *RangeModRangeSum) Set(pos, value int) {
	if pos < 0 || pos >= len(r.nums) {
		return
	}
	pre := r.nums[pos]
	if pre == value {
		return
	}
	r.nums[pos] = value
	r.rebuild(r.belong[pos])
}

func (r *RangeModRangeSum) rebuild(bid int) {
	r.blockSum[bid] = 0
	r.blockMax[bid] = 0
	for i := r.blockStart[bid]; i < r.blockEnd[bid]; i++ {
		r.blockSum[bid] += r.nums[i]
		r.blockMax[bid] = max(r.blockMax[bid], r.nums[i])
	}
}

// blockSize = int(math.Sqrt(float64(len(nums)))+1)
func UseBlock(n int, blockSize int) struct {
	belong     []int // 下标所属的块.
	blockStart []int // 每个块的起始下标(包含).
	blockEnd   []int // 每个块的结束下标(不包含).
	blockCount int   // 块的数量.
} {
	blockCount := 1 + (n / blockSize)
	blockStart := make([]int, blockCount)
	blockEnd := make([]int, blockCount)
	belong := make([]int, n)
	for i := 0; i < blockCount; i++ {
		blockStart[i] = i * blockSize
		tmp := (i + 1) * blockSize
		if tmp > n {
			tmp = n
		}
		blockEnd[i] = tmp
	}
	for i := 0; i < n; i++ {
		belong[i] = i / blockSize
	}

	return struct {
		belong     []int
		blockStart []int
		blockEnd   []int
		blockCount int
	}{belong, blockStart, blockEnd, blockCount}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
